internal/controllers/stats.go

```diff
@@ -64,20 +63,90 @@ func (c *ApiController) StatisticsHandlerAllStatsForLoadbalancer(w http.Response
 		return
 	}
 
-	statistics, err := payout.GetStatsForPayout(c.repositories, getNow(), true)
+	totalRewardAsFloat, err := getTotalRewardFromRequest(r)
+	if err != nil {
+		log.Error(err)
+		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+	}
+
+	timestamp := getNow()
+	statistics, err := stats.CalculateStatisticsFromLastPayout(c.repositories, timestamp)
 	if err != nil {
 		log.Errorf("Failed to calculate statistics, because %v", err)
 		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
 		return
 	}
 
+	err = c.repositories.PayoutRepo.Save(&models.Payout{
+		Timestamp:      timestamp,
+		PaymentDetails: statistics,
+		LbFee:          totalRewardAsFloat * float64(configuration.Config.Fee),
+	})
+	if err != nil {
+		log.Errorf("Failed to save payout, because %v", err)
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	feesToNodes := payout.CalculatePayoutDistributionByNode(
+		statistics,
+		totalRewardAsFloat,
+		payout.LoadBalancerDistributionConfiguration{
+			FeePercentage:       float64(configuration.Config.Fee),
+			DifferentFeeAddress: false,
+		},
+	)
+	for nodeId, amount := range feesToNodes {
+		err := c.repositories.FeeRepo.RecordNewFee(nodeId, amount.Int64())
+		if err != nil {
+			log.Errorf("Failed to save fee for node %s, because %v", nodeId, err)
+		}
+	}
+
 	w.Header().Set("Content-Type", "application/json")
 	_ = json.NewEncoder(w).Encode(LoadbalancerStatsResponse{
 		Stats: statistics,
 		Fee:   configuration.Config.Fee,
 	})
 }
 
+func verifySignatureInHeader(r *http.Request, privateKey string) (bool, int, error) {
```

Comment:
Can you move this logic to some middleware?

Reply:
I could, but then it would need additional refactoring so the private key is exposed when creating routes and adding this middleware. Currently private key is property of `ApiController`
I agree with you that this kind of code should be moved to middleware, but as this verification is only used for one route (and we are pretty sure that is the only use-case) I think it is not the end of the world to keep it as it is.

I will check how much refactoring is needed and make decision based on that if that is ok with you?

Reply:
:+1:

Reply:
It shouldn't be much of refacatoring, so I will move it to separate middleware.